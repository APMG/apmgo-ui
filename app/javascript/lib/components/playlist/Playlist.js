// @flow
import * as React from 'react'
import { connect } from 'react-redux'
import { Preview } from 'react-dnd-multi-backend'

import ItemSlot from './ItemSlot'
import PlaylistItem, { PlaylistItemDragPreview } from './PlaylistItem'
import VerticalScrollingList from './VerticalScrollingList'
import type { PlaylistItemType } from '../../redux/types'
import AudioPlayer from '../player/AudioPlayer'
import { playlistItemMoved } from '../../redux/data'

import './Playlist.scss'

type PlaylistProps = {
  playlist: Array<PlaylistItemType>,
  activeItem?: PlaylistItemType,
  playlistItemMoved: (item: PlaylistItemType, newPosition: number) => {}
}

function generatePreview (type: string, item: any, style: any) {
  return <PlaylistItemDragPreview
    item={item.item}
    style={style}
  />
}

export class PlaylistPresenter extends React.Component {
  props: PlaylistProps

  render () {
    if (!this.props.playlist || !this.props.playlist.length) {
      return <h1>Loading ...</h1>
    }
    return (
      <div>
        <AudioPlayer item={this.props.activeItem} />
        <VerticalScrollingList>
          {this.props.playlist
            .filter(item => !item.attributes.finished)
            .map((item, i) =>
              <ItemSlot key={item.attributes.audio_identifier} index={i}>
                <PlaylistItem
                  playlistItemMoved={this.props.playlistItemMoved}
                  item={item}
                  index={i}
                />
              </ItemSlot>
            )
          }
          <Preview generator={generatePreview} />
        </VerticalScrollingList>
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  let activeItem = ownProps.playlist.find(item => item.id === state.audioPlayer.currentTrackId)

  if (!activeItem) {
    activeItem = ownProps.playlist.find((item: PlaylistItemType) => {
      return !item.attributes.finished
    })
  }

  return {
    activeItem: Object.assign({}, activeItem)
  }
}

const mapDispatchToProps = (dispatch: (action: any) => {}, ownProps) => {
  return {
    playlistItemMoved: (item: PlaylistItemType, newPosition: number) => {
      let newAfter = ownProps.playlist[newPosition - 1]

      newAfter = newAfter ? newAfter.id : null

      dispatch(playlistItemMoved(item, newAfter))
    }
  }
}

const PlaylistContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(PlaylistPresenter)

export default PlaylistContainer
