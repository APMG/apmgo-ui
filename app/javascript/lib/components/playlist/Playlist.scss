@import '../../styles/helpers';

.list {
  margin-bottom: 20px;
  position: relative;
  z-index: $z-playlist;

  > * {
    padding-bottom: 10px;
  }
}
