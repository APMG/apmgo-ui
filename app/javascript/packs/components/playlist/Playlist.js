// @flow
import * as React from 'react'
import { connect } from 'react-redux'
import PlaylistItem from './PlaylistItem'
import List from 'material-ui/List'
import type { PlaylistItemType } from '../../redux/types'
import AudioPlayer from '../player/AudioPlayer'
import Async from 'react-code-splitting'

type PlaylistProps = {
  data: Array<PlaylistItemType>,
  activeItem?: PlaylistItemType
}

interface TimelineListState {}

export class PlaylistPresenter extends React.Component {

  constructor(props: PlaylistProps) {
    super(props)
  }

  render () {
    if (!this.props.data || !this.props.data.length) {
      return <h1>Loading ...</h1>
    }
    return (
      <div>
        <AudioPlayer item={this.props.activeItem}/>
        <List>
          {this.props.data.map((item, i) =>
            <PlaylistItem item={item} key={i} />
          )}
        </List>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  let activeItem = state.data.data.find(item => item.id === state.audioPlayer.currentTrackId)
  return {
    data: state.data.data,
    activeItem: Object.assign({}, activeItem)
  }
}

const PlaylistContainer = connect(
  mapStateToProps
)(PlaylistPresenter)

export default PlaylistContainer
