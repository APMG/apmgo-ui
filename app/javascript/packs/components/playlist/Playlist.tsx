import * as React from 'react'
import { connect } from 'react-redux'
import PlaylistItem from './PlaylistItem'
import List from 'material-ui/List'
import ListItem from 'material-ui/List/ListItem'
import { PlaylistItemType } from '../../redux/types'

interface TimelineListProps {
  data: Array<PlaylistItemType>
}

interface TimelineListState {}

class Playlist extends React.Component<TimelineListProps, TimelineListState> {
  constructor(props) {
    super(props)
  }

  render () {
    if (!this.props.data) {
      return <h1>Loading ...</h1>
    }
    return (
      <List>
        {this.props.data.map((item, i) =>
          <PlaylistItem item={item} key={i} />
        )}
      </List>
    )
  }
}

const mapStateToProps = (state) => {
  let playlistItems = state.data.data

  return { data: playlistItems }
}

const PlaylistContainer = connect(
  mapStateToProps
)(Playlist)

export default PlaylistContainer
