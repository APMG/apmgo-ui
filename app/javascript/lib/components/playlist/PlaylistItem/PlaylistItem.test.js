import React from 'react'
import { PlaylistPresenter } from '../Playlist'
import { getSnapshotJson } from '../../../__tests__/testHelpers'

jest.mock('../../../service/auth-layer')

describe('Playlist Component', () => {
  describe('Presenter', () => {
    it('Renders', () => {
      let tree = getSnapshotJson(<PlaylistPresenter />)
      expect(tree).toMatchSnapshot()
    })
    it('Renders With Empty Data', () => {
      let tree = getSnapshotJson(<PlaylistPresenter />, {data: []})
      expect(tree).toMatchSnapshot()
    })
  })
})
