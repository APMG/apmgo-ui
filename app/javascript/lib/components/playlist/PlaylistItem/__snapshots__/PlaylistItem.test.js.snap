// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist Component Presenter Renders 1`] = `
<h1>
  Loading ...
</h1>
`;

exports[`Playlist Component Presenter Renders With Empty Data 1`] = `
<h1>
  Loading ...
</h1>
`;
